In [1]:
import heapq
from collections import Counter

In [2]:
#FILE = "example.txt"
FILE = "input.txt"

In [3]:
inputs = [
    list(line.split()) for line in open(FILE)
]

In [4]:
ranges = []
while inputs:
    line = inputs.pop(0)
    if not line:
        break
    ranges.extend(line)
ranges = [(int(i.split('-')[0]),int(i.split('-')[1])) for i in ranges]

In [5]:
ranges = sorted(sorted(ranges, key=lambda x: x[1]), key=lambda x: x[0])

In [6]:
len(ranges)

185

In [7]:
# ranges = [(0,2),(1,10),(12,20),(13,15),(13,16),(25,30),(25,26),(26,26),(31,31),(31,32),(31,33),(32,40),(39,41),(45,45),(45,45)]
# ranges = sorted(sorted(ranges, key=lambda x: x[1]), key=lambda x: x[0])
idx = 0
end_ranges = []
while idx < len(ranges):
    cur_range = ranges[idx]
    i = idx + 1
    end_candidate = cur_range[1]
    while i < len(ranges) and (
        (ranges[i][0] <= end_candidate or ranges[i][1] <= end_candidate)
    ):
        end_candidate = max(ranges[i][1],end_candidate)
        i += 1
    end_ranges.append((cur_range[0], end_candidate))
    idx = i

In [8]:
fresh_ids = 0
for start,end in end_ranges:
    fresh_ids += end-start+1

In [9]:
fresh_ids

347338785050515

366571009196172 was too high, I forgot to also sort by end

337215458895890 was too low

337215458895905 was still too low, even after I compared ranges between +1/-1

347338785050530 was wrong, after I fixed the correct sorting order

347338785050515 after removing the +1/-1, it was the right answer